### PR TITLE
Implement 0.8 spec

### LINKED ISSUE #47: Support UCAN 0.8.x
The spec is live!

https://github.com/ucan-wg/spec

### PR DESCRIPTION
**Pull request by icidasset**
This PR implements the [0.8.1 spec](https://github.com/ucan-wg/spec) and thus closes #47.

The biggest change involves attenuations. I've taken the capability types, revised them to the new spec, taken them out of the `types.ts` module and split them up in several parts under `src/capabilities/`, plus the top level `src/capabilities.ts` module. Each module in there has the types, utility functions, encoders/decoders, type checking functions, etc.

Notes:
- Introduced semver version parsing in the compatibility module
- Legacy UCAN version `0.0.1` has been changed to `0.3.0`
- Categorised things in the types module
- Moved WNFS things to tests, should be moved to webnative in the future.
- Moved `tests/emailCapabilities.ts` to `tests/capabilities/email.ts`

Todo/Questions:
- [x] Do we need additional utility functions for the capabilities?
- [x] Do we need more spec enforcing? (ie. such as enforcing the URI format for capability resource pointers)
- [x] Does the spec enforcing cause any issues anywhere? (eg. check encoding/decoding of URI format of resource pointers)
- [x] Did I miss anything in the spec? (go over spec once more)

### CODE REVIEW

File: src/capability.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  }
+}
+
+
+export function prf(selector: Superuser | number, ability: Ability): Capability {
+  return {
+    with: resourcePointer.prf(selector),
+    can: ability
+  }
+}
```
Comment (icidasset) [inline src/capability.ts:74]: Should we have functions for `wnfs` and `email` too? I realise they're not part of the spec, but could be helpful?

**Comment by icidasset** [comment]
I think this PR also partially implements #43? It's got some utility functions for creating `prf` capabilities and for superuser`*` resources/abilities.

> In capability checking it essentially allows you to "forward" who the "originator" is for all capabilities without having to mention concrete capabilities.

@matheus23 Do you think that after my 0.8 changes we need to adjust the `CapabilitySemantics` parsing/delegating?

File: README.md [hunk 2]
```diff
@@ -44,41 +47,52 @@ Read more in the whitepaper: https://whitepaper.fission.codes/access-control/uca
...
+`capabilities` is an array of resource pointers and abilities:
 ```ts
 {
-  $TYPE: $IDENTIFIER,
-  "cap": $CAPABILITY
+  // `with` is a resource pointer in the form of a URI, which has a `scheme` and `hierPart`.
+  // → "mailto:boris@fission.codes"
+  with: { scheme: "mailto", hierPart: "boris@fission.codes" },
+
+  // `can` is an ability, which always has a namespace and optional parts.
```
Comment (bgins) [inline README.md:88]: Might be worth changing "optional parts" to "optional segments" to match the `segments` param.

File: src/token.ts [hunk 3]
```diff
@@ -200,48 +209,69 @@ export function encodePayload(payload: UcanPayload): string {
...
- * @param encodedUcanHeader The encoded UCAN header.
+ * @param encodedUcan The encoded UCAN.
  */
-export function parseHeader(encodedUcanHeader: string): unknown {
-  let decodedUcanHeader: string
+export function parse(encodedUcan: string): UcanParts {
+  const [ encodedHeader, encodedPayload, signature ] = encodedUcan.split(".")
+
+  if (encodedHeader == null || encodedPayload == null || signature == null) {
+    throw new Error(`Can't parse UCAN: ${encodedUcan}: Expected JWT format: 3 dot-separated base64url-encoded values.`)
```
Comment (bgins) [inline src/token.ts:220]: Great error message! 💪

**Review by bgins — APPROVED** [review]
Looks great! Love how the customizable capability semantics have come out here. 💯 🎉

File: src/token.ts (hunk 3 above)
Comment (icidasset) [reply src/token.ts:220]: Wasn't me, I just moved it 😄

**Comment by icidasset** [comment]
I'll wait for @matheus23 to give this a look as well.

**Pull request merged**